Include BoundingBox and OrientedBoundingBox types

gustavoneves12 opened the pull request:
This PR includes two data types to represent axis-aligned and oriented bounding boxes with their uncertainty associated. These two data types could be used on object recognition applications.

doudou commented on src/samples/BoundingBox.hpp:
```diff
@@ -0,0 +1,42 @@
+
+    BoundingBox()
+        : time(),
+          position(base::Vector3d(0, 0, 0)),
```
The convention so far for pre-initialized types is to always use `base::unknown` (a.k.a. NAN)

doudou commented on src/samples/BoundingBox.hpp:
```diff
@@ -0,0 +1,42 @@
+        : time(),
+          position(base::Vector3d(0, 0, 0)),
+          dimension(base::Vector3d(0, 0, 0)),
+          cov_position(Matrix3d::Ones() * NaN<double>()),
```
Please use `unknown<double>` to be explicit on the semantics.

doudou commented on src/samples/OrientedBoundingBox.hpp:
```diff
@@ -0,0 +1,30 @@
+namespace base {
+namespace samples {
+/**
+ * Oriented bounding box (OBB) representation with associated uncertainty
+ * information.
```
Please document the semantics of the rotation. Is it applied in whatever frame the box itself is defined ? Is it a rotation around the center position ? (I'm guessing the latter, but please document it)

gustavoneves12 replied:
It is a rotation around the center position. I am going to fix it.

doudou commented on src/samples/OrientedBoundingBox.hpp:
```diff
@@ -0,0 +1,30 @@
+
+    OrientedBoundingBox()
+        : BoundingBox(),
+          orientation(base::Quaterniond::Identity()),
+          cov_orientation(Matrix3d::Ones() * NaN<double>()) {
```
Same comment about `unknown`

doudou commented on test/test.cpp:
```diff
@@ -1264,6 +1266,37 @@ BOOST_AUTO_TEST_CASE( pose_with_covariance )
+
+    for (int i = 0; i < 3; ++i) {
+        for (int j = 0; j < 3; ++j) {
+            BOOST_CHECK(base::isNaN(bb.cov_position(i, j)));
+            BOOST_CHECK(base::isNaN(bb.cov_dimension(i, j)));
+        }
```
Use `isUnknown` instead of `isNaN`

gustavoneves12 commented:
@doudou I've changed the code following your suggestion. Could you check if everything is correct? Thank you in advance.

gustavoneves12 commented:
@doudou I have leveraged this PR and also included a [commit](https://github.com/rock-core/base-types/pull/147/commits/66ef3bc7dbb9732481c3c80a97e0f8f026c68378) which initializes the `base::Waypoint` with unknown values. Could you please review it?

Thanks in advance

g-arjones commented:
@doudou Ping?

doudou commented on src/samples/BoundingBox.cpp:
```diff
@@ -0,0 +1,60 @@
+}
+
+bool BoundingBox::hasValidPosition() const {
+    return position.allFinite();
```
TIL about `allFinite`

doudou commented on src/samples/OrientedBoundingBox.hpp:
```diff
@@ -0,0 +1,86 @@
+struct OrientedBoundingBox : public BoundingBox {
+    OrientedBoundingBox();
+
+    OrientedBoundingBox(const base::Time& time,
+                        const base::Vector3d& position);
+
+    OrientedBoundingBox(const base::Time& time,
+                        const base::Vector3d& position,
+                        const base::Vector3d& dimension);
+
+    OrientedBoundingBox(const base::Time& time,
+                        const base::Vector3d& position,
+                        const base::Vector3d& dimension,
+                        const base::Vector4d& orientation);
+
+    OrientedBoundingBox(const base::Time& time,
+                        const base::Vector3d& position,
+                        const base::Vector3d& dimension,
+                        const base::Orientation& orientation);
```
Any reason why you don't use default arguments instead of overloads, which would allow to have a single constructor ... (Just remember to make it explicit).

gustavoneves12 replied:
No reason that I can remember now. I am going to change for default arguments.

gustavoneves12 replied:
I've changed as you suggested. I replaced the multiple overloaded constructors by a single one with default arguments.

doudou commented:
> @doudou Ping?

Sorry. I've been staying away from GitHub, I have a hard time avoiding that it eats my time. I'll keep a daily schedule looking at stuff from now on.

Sorry again.

gustavoneves12 commented:
>@gustavoneves12 the change on Waypoint is not backward compatible. This is a rather old type, and there certainly is old code that relies on the current initialization behavior. Please don't include that change here.

@doudou I have changed the waypoint data type to make it backward compatible. In this modification, I have included a new constructor (`Waypoint(position, heading = 0`) and added the `base::unknown<Waypoint>()` function to initialize it with unknown values. I think those modifications will not affect existent old codes that rely on this data type.

gustavoneves12 commented on src/samples/OrientedBoundingBox.hpp:
```diff
@@ -0,0 +1,72 @@
+    OrientedBoundingBox(const Time& time,
+                        const Vector3d& position,
+                        const Vector3d& dimension,
+                        const Vector4d& orientation);
```
I've removed the constructor with `base::Vector4d`.
